src/main/java/com/google/sps/servlets/utils/YouTubeCommentRetriever.java

```diff
@@ -0,0 +1,100 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.sps.servlets.utils;
+
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.youtube.YouTube;
+import com.google.api.services.youtube.model.*;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.List;
+
+/*
+ * Class to retrieve YouTube comments from a designated URL with certain parameters
+ */
+public class YouTubeCommentRetriever {
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+  // Limit of comments that can be obtained in one request
+  private static final Long COMMENT_LIMIT = 100L;
+  // Parameters required by YouTube API to retrieve the comment threads
+  private static final String SNIPPET_PARAMETERS = "snippet,replies";
+  private static final String ORDER_PARAMETER = "relevance";
+
+  private static final String APPLICATION_NAME = "SAY";
+  // TODO: have dev key come from centralized location, rather than being hard-coded.
+  private static final String DEVELOPER_KEY = "AIzaSyDYfjWcy1hEe0V7AyaYzgIQm_rT-9XbiGs";
+
+  public static List<CommentThread> retrieveComments(String url, int maxComments) {
+    String nextPageToken = "";
+    int currentOverallComments = 0;
+    long commentQueryLimit = 0;
+    ArrayList<CommentThread> allComments = new ArrayList<CommentThread>();
+    try {
+      do {
+        commentQueryLimit = Math.min(COMMENT_LIMIT, maxComments - currentOverallComments);
+        YouTube.CommentThreads.List commentRequest = generateYouTubeRequest(url, commentQueryLimit);
+        if (nextPageToken != null && nextPageToken != "") {
+          commentRequest.setPageToken(nextPageToken);
+        }
+        CommentThreadListResponse commentResponse = commentRequest.execute();
+        nextPageToken = commentResponse.getNextPageToken();
+        currentOverallComments += COMMENT_LIMIT;
+        for (CommentThread thread : commentResponse.getItems()) {
+          // Extract relevant comment from CommentThreadListResponse
+          allComments.add(thread);
+        }
+        // Todo: consolidate comments into large list
+      } while (nextPageToken != null
+          && nextPageToken != ""
+          && currentOverallComments < maxComments);
```

Comment:
would suggest to refactor this to a direct while loop or for loop

Reply:
This already is a while loop. What do you mean by direct while loop?